src/modules/partition/core/PartitionLayout.h

```diff
@@ -0,0 +1,70 @@
+/* === This file is part of Calamares - <https://github.com/calamares> ===
+ *
+ *   Copyright 2018, Collabora Ltd
+ *
+ *   Calamares is free software: you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation, either version 3 of the License, or
+ *   (at your option) any later version.
+ *
+ *   Calamares is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with Calamares. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef PARTITIONLAYOUT_H
+#define PARTITIONLAYOUT_H
+
+#include "Typedefs.h"
+
+// KPMcore
+#include <kpmcore/core/partitiontable.h>
+
+// Qt
+#include <QList>
+#include <QObject>
+
+class Partition;
+
+class PartitionLayout
+{
+public:
+
+    enum SizeUnit
+    {
+        Percent = 0,
+        Byte,
+        KiB,
+        MiB,
+        GiB
+    };
+
+    struct PartitionEntry
+    {
+        QString partLabel;
+        QString partMountPoint;
+        int partFileSystem;
+        double partSize;
+        SizeUnit partSizeUnit;
+        double partMinSize;
+        SizeUnit partMinSizeUnit;
+    };
+
+    PartitionLayout();
+    PartitionLayout( PartitionEntry entry );
+    PartitionLayout( const PartitionLayout& layout );
+    ~PartitionLayout();
+
+    void addEntry( PartitionEntry entry );
+    void addEntry( QString mountPoint, QString size, QString min = "" );
```

Comment:
`min = QString()` is more idiomatic nowadays. Also one typically passes `const QString&`, even though they're implicitly shared and this doesn't make a lot of difference performance-wise.

Reply:
.. I might have some stylistic comments, but overall at-a-quick-glance this looks amazingly well-done.

Reply:
The point of the const reference is to avoid pointless atomic reference count operations.

And `min = QString()` is not just "more idiomatic", but also faster than `min = ""` that explicitly constructs a non-null empty QString.

Reply:
OK, will fix it, thanks for the explanations.
(my C++ is a bit rusty, I've been doing C-only work for the last 10 years or so)

I'll leave you some time to review the whole thing and will submit a batch of fixes to take everything into account, if you're fine with it.